### PR TITLE
llvm.lround: Update verifier to validate support of vector types.

### PR DESCRIPTION
**Pull request by sgundapa**
Both IRVerifier and Machine Verifier are updated

### CODE REVIEW

File: llvm/unittests/IR/IntrinsicsTest.cpp [hunk 1]
```diff
@@ -106,4 +107,67 @@ TEST_F(IntrinsicsTest, InstrProfInheritance) {
     EXPECT_TRUE(Checker(*Intr));
   }
 }
+
+TEST(IntrinsicVerifierTest, LRound) {
```
Comment (arsenm) [inline llvm/unittests/IR/IntrinsicsTest.cpp:111]: Would be easier to just write regular IR in test/Assembler

**Review by arsenm — COMMENTED** [review]
Should update the LangRef.

Don't see why we need this, but I guess it improves consistency

File: llvm/unittests/IR/IntrinsicsTest.cpp (hunk 1 above)
Comment (sgundapa) [reply llvm/unittests/IR/IntrinsicsTest.cpp:111]: I am planning to extend the same to lrint as well.

**Comment by sgundapa** [comment]
I have updated the document for consistency

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 2]
```diff
@@ -2032,7 +2032,20 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
   }
   case TargetOpcode::G_LLROUND:
   case TargetOpcode::G_LROUND: {
-    verifyAllRegOpsScalar(*MI, *MRI);
+    LLT DstTy = MRI->getType(MI->getOperand(0).getReg());
+    LLT SrcTy = MRI->getType(MI->getOperand(1).getReg());
+    if (!DstTy.isValid() || !SrcTy.isValid())
+      break;
+    if (SrcTy.isPointer() || DstTy.isPointer()) {
+      std::string Op = SrcTy.isPointer() ? "Source" : "Destination";
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineVerifier.cpp:2040]: This can be a StringRef

**Comment by sgundapa** [comment]
The scalarizatio of vector input happens through this patch : https://github.com/llvm/llvm-project/pull/101054

File: llvm/unittests/IR/IntrinsicsTest.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/unittests/IR/IntrinsicsTest.cpp:111]: Same? Can just write an assembler test?

**Review by arsenm — COMMENTED** [review]
Still not sure why we need this

File: llvm/test/Assembler/lround_vec_intrinsic.ll [hunk 3]
```diff
@@ -0,0 +1,29 @@
+; Validate that vector types are accepted for llvm.lround/llvm.llround intrinsic
+; RUN: llvm-as < %s -disable-output 2>&1| FileCheck -allow-empty %s
+
+; CHECK-NOT:assembly parsed, but does not verify as correct
+; CHECK-NOT:Intrinsic does not support vectors
```
Comment (arsenm) [inline llvm/test/Assembler/lround_vec_intrinsic.ll]: Check the actual output. not checks are fragile, especially of error output

File: llvm/test/Assembler/lround_vec_intrinsic.ll [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+define <2 x i32> @intrinsic_lround_v2i32_v2f32(<2 x float> %arg) {
+entry:
+  %0 = tail call <2 x i32> @llvm.lround.v2i32.v2f32(<2 x float> %arg)
+  ret <2 x i32> %0
```
Comment (arsenm) [inline llvm/test/Assembler/lround_vec_intrinsic.ll:10]:
```suggestion
  %ret = tail call <2 x i32> @llvm.lround.v2i32.v2f32(<2 x float> %arg)
  ret <2 x i32> %ret
```

File: llvm/test/Assembler/lround_vec_intrinsic.ll [hunk 5]
```diff
@@ -0,0 +1,29 @@
+; Validate that vector types are accepted for llvm.lround/llvm.llround intrinsic
+; RUN: llvm-as < %s -disable-output 2>&1| FileCheck -allow-empty %s
+
```
Comment (arsenm) [inline llvm/test/Assembler/lround_vec_intrinsic.ll:3]: Rename test to just lround?

File: llvm/test/Assembler/lround_vec_intrinsic.ll (hunk 3 above)
Comment (sgundapa) [reply llvm/test/Assembler/lround_vec_intrinsic.ll]: Sure Matt.

File: llvm/lib/IR/Verifier.cpp [hunk 6]
```diff
@@ -5963,8 +5963,21 @@ void Verifier::visitIntrinsicCall(Intrinsic::ID ID, CallBase &Call) {
...
+        "llvm.lround, llvm.llround: argument must be floating-point or vector "
+        "of floating-points, and result must be integer or vector of integers",
+        &Call);
+    Check(
+        ValTy->isVectorTy() == ResultTy->isVectorTy(),
+        "llvm.lround, llvm.llround: argument and result disagree on vector use",
+        &Call);
+    if (ValTy->isVectorTy()) {
+      Check(cast<VectorType>(ValTy)->getElementCount() ==
+                cast<VectorType>(ResultTy)->getElementCount(),
```
Comment (arsenm) [inline llvm/lib/IR/Verifier.cpp:5977]: dyn_cast instead of isVectorTy + 2 x cast

File: llvm/test/Assembler/lround.ll [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  ;CHECK: %res = tail call <2 x i64> @llvm.lround.v2i64.v2f32(<2 x float> %arg)
+  %res = tail call <2 x i64> @llvm.lround.v2i64.v2f32(<2 x float> %arg)
+  ret <2 x i64> %res
+}
+
+define <2 x i64> @intrinsic_llround_v2i64_v2f32(<2 x float> %arg) {
+entry:
+  ;CHECK: %res = tail call <2 x i64> @llvm.llround.v2i64.v2f32(<2 x float> %arg)
+  %res = tail call <2 x i64> @llvm.llround.v2i64.v2f32(<2 x float> %arg)
+  ret <2 x i64> %res
```
Comment (arsenm) [inline llvm/test/Assembler/lround.ll:29]:
```suggestion
define <2 x i32> @intrinsic_lround_v2i32_v2f32(<2 x float> %arg) {
  ;CHECK: %res = tail call <2 x i32> @llvm.lround.v2i32.v2f32(<2 x float> %arg)
  %res = tail call <2 x i32> @llvm.lround.v2i32.v2f32(<2 x float> %arg)
  ret <2 x i32> %res
}

define <2 x i32> @intrinsic_llround_v2i32_v2f32(<2 x float> %arg) {
  ;CHECK: %res = tail call <2 x i32> @llvm.llround.v2i32.v2f32(<2 x float> %arg)
  %res = tail call <2 x i32> @llvm.llround.v2i32.v2f32(<2 x float> %arg)
  ret <2 x i32> %res
}

define <2 x i64> @intrinsic_lround_v2i64_v2f32(<2 x float> %arg) {
  ;CHECK: %res = tail call <2 x i64> @llvm.lround.v2i64.v2f32(<2 x float> %arg)
  %res = tail call <2 x i64> @llvm.lround.v2i64.v2f32(<2 x float> %arg)
  ret <2 x i64> %res
}

define <2 x i64> @intrinsic_llround_v2i64_v2f32(<2 x float> %arg) {
  ;CHECK: %res = tail call <2 x i64> @llvm.llround.v2i64.v2f32(<2 x float> %arg)
  %res = tail call <2 x i64> @llvm.llround.v2i64.v2f32(<2 x float> %arg)
  ret <2 x i64> %res
```

**Review by arsenm — APPROVED** [review]

**Pull request merged**